walkthrough-w-5 (tdd) Add calendar service to handle incoming meetings scheduling for users

dmosc opened the pull request:
feature (FindMeetingQuery.java) Construct FindMeetingQuery.java functionality to handle meeting scheduling.

+ Added Comparator function to Event.java to sort Event objects by start date.

+ Fill query function in FindMeetingQuery.java to search for potential timeslots to
fit an incoming meeting request.

+ Added additional constructor to MeetingRequest.java to include optional_attendes
if necessary to build test cases that included optional attendees.

+ Added requested test cases in FindMeetingQueryTest.java to test for different
scenarios mentioned in the walkthrough.

dmosc committed:
feature (FindMeetingQuery.java) Construct FindMeetingQuery.java functionality to handle meeting scheduling.

+ Added Comparator function to Event.java to sort Event objects by start date.

+ Fill query function in FindMeetingQuery.java to search for potential timeslots to
fit an incoming meeting request.

+ Added additional constructor to MeetingRequest.java to include optional_attendes
if necessary to build test cases that included optional attendees.

+ Added requested test cases in FindMeetingQueryTest.java to test for different
scenarios mentioned in the walkthrough.

*** No breaking changes ***

Domene99 commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -15,9 +15,91 @@
+        ArrayList<TimeRange> availableTimeRanges = new ArrayList<>();
+        ArrayList<Event> eventsToSet = new ArrayList<>(events);
+
+        Collections.sort(eventsToSet, Event.ORDER_BY_START); // N * log N
```
I don't think you need to sort for your solution

dmosc replied:
That makes sense lol.

Will fix!

ssmall commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/Event.java:
```diff
@@ -53,6 +54,13 @@ public Event(String title, TimeRange when, Collection<String> attendees) {
     this.attendees.addAll(attendees);
   }
 
+  public static final Comparator<Event> ORDER_BY_START = new Comparator<Event>() {
+    @Override
+    public int compare(Event a, Event b) {
+      return Integer.compare(a.when.start(), b.when.start());
+    }
+  };
```
nit: this could be expressed in fewer lines by using a lambda expression

dmosc replied:
Yeah that's true. But I just realized that it is not necessary anymore to sort the Events for my implementation. So I'll just go ahead and remove this Comparator since I'm not using it anymore. :)

ssmall commented on walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:
```diff
@@ -121,6 +124,63 @@ public void everyAttendeeIsConsidered() {
 
+  @Test
+  public void everyAttendeeIsConsideredWithOptionalAttendeeAllDayEvent() {
+    // Have each person have different events. We should see two options because each person has
+    // split the restricted times.
+    //
+    // Events  :       |--A--|     |--B--|
+    // Day     : |-----------------------------|
+    // Options : |--1--|     |--2--|     |--3--|
```
Comment text says "we should see two options" but the illustration shows three options

ssmall commented on walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:
```diff
@@ -121,6 +124,63 @@ public void everyAttendeeIsConsidered() {
+            Arrays.asList(PERSON_A)),
+        new Event("Event 2", TimeRange.fromStartDuration(TIME_0900AM, DURATION_30_MINUTES),
+            Arrays.asList(PERSON_B)),
+        new Event("Event 3", TimeRange.WHOLE_DAY,
+            Arrays.asList(PERSON_C)));
```
Person C does not appear to be represented in the illustration at the top of the test

ssmall commented on walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:
```diff
@@ -121,6 +124,63 @@ public void everyAttendeeIsConsidered() {
+
+  @Test
+  public void everyAttendeeIsConsideredWithOptionalAttendeeOverlappingEvent() {
+    // Have each person have different events. We should see two options because each person has
+    // split the restricted times.
+    //
+    // Events  :       |--A--|     |--B--|
+    // Day     : |-----------------------------|
+    // Options : |--1--|     |--2--|     |--3--|
+
+    Collection<Event> events = Arrays.asList(
+        new Event("Event 1", TimeRange.fromStartDuration(TIME_0800AM, DURATION_30_MINUTES),
+            Arrays.asList(PERSON_A)),
+        new Event("Event 2", TimeRange.fromStartDuration(TIME_0900AM, DURATION_30_MINUTES),
+            Arrays.asList(PERSON_B)),
+        new Event("Event 3", TimeRange.fromStartDuration(TIME_0830AM, DURATION_30_MINUTES),
+            Arrays.asList(PERSON_C)));
```
description and illustration does not match the test inputs

ssmall commented on walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:
```diff
@@ -223,6 +283,32 @@ public void justEnoughRoom() {
 
+  @Test
+  public void justEnoughRoomWithOptionalAttendee() {
+    // Have one person, but make it so that there is just enough room at one point in the day to
+    // have the meeting.
+    //
+    // Events  : |--A--|     |----A----|
+    // Day     : |---------------------|
+    // Options :       |-----|
+
+    Collection<Event> events = Arrays.asList(
+        new Event("Event 1", TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TIME_0830AM, false),
+            Arrays.asList(PERSON_A)),
+        new Event("Event 2", TimeRange.fromStartEnd(TIME_0900AM, TimeRange.END_OF_DAY, true),
+            Arrays.asList(PERSON_A)),
+        new Event("Event 3", TimeRange.fromStartEnd(TIME_0830AM, TIME_0845AM, true),
+            Arrays.asList(PERSON_B)));
```
description and illustration does not match test inputs

ssmall commented on walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:
```diff
@@ -270,5 +356,43 @@ public void notEnoughRoom() {
   }
+
+  @Test
+  public void twoOptionalAttendeesWithSeveralGaps() {
+    Collection<Event> events = Arrays.asList(
+        new Event("Event 1", TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TIME_0830AM, false),
+            Arrays.asList(PERSON_A)),
+        new Event("Event 2", TimeRange.fromStartEnd(TIME_1000AM, TimeRange.END_OF_DAY, false),
+            Arrays.asList(PERSON_A)),
+        new Event("Event 3", TimeRange.fromStartEnd(TIME_0830AM, TIME_0900AM, false),
+            Arrays.asList(PERSON_B)),
+        new Event("Event 4", TimeRange.fromStartEnd(TIME_0930AM, TimeRange.END_OF_DAY, true),
+            Arrays.asList(PERSON_B)));
```
It would be helpful to add an illustration as in the other tests

ssmall commented on walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:
```diff
@@ -270,5 +356,43 @@ public void notEnoughRoom() {
+  }
+
+  @Test
+  public void twoOptionalAttendeesWithNoGaps() {
+    Collection<Event> events = Arrays.asList(
+        new Event("Event 1", TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TimeRange.END_OF_DAY, true),
+            Arrays.asList(PERSON_A)),
+        new Event("Event 2", TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TimeRange.END_OF_DAY, true),
+            Arrays.asList(PERSON_B)));
```
It would be helpful to add an illustration as in the other tests

ssmall commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -15,9 +15,89 @@
-    throw new UnsupportedOperationException("TODO: Implement this method.");
-  }
-}
+    public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
```
Could you please include a javadoc comment with the following information:
1. A high-level description of the algorithm
2. The worst-case (big-O) runtime and space complexity, expressed as a function of the inputs

dmosc replied:
Sure!

ssmall reviewed (commented):
Looks good! See comments below.

dmosc committed:
nitfix (Comments) Add comments to tests and Javadoc description of FindMeetingQuery

- Removed Comparator function from Event.java since it is not necessary to sort the events.

+ Added Javadoc comment in FindMeetingQuery.java to describe and analyze algorithm complexity.

+ Add relevant comments to each test case in FindMeetingQueryTest.java to better describe what each test
works for.

*** No breaking changes ***

anniefu commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -15,9 +15,89 @@
+            /* Verify that at least one of the request attendees
+            is scheduled for the current event. */
+            Set<String> eventAttendees = event.getAttendees();
+            Boolean isAnyoneBusy = !Collections.disjoint(eventAttendees, attendees); // O + P
```
[nit] maybe rename to `isAnyMandatoryBusy`, otherwise "Anyone" could refer to both mandatory + optional.

anniefu commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -15,9 +15,89 @@
+            /* Mark available spaces as busy since there's at
+                least one person from the request that isn't available */
+            if (isAnyoneBusy) {
+                Arrays.fill(minutesNoOptAttendees, start, end, 0); // end - start ops.
```
This can be refactored to save a couple of checks:
```
if (isAnyoneBusy) {
    Arrays.fill(minutesNoOptAttendees...)
    Arrays.fill(minutesOptAttendees...)
} else if (isAnyOptionalBusy) {
    Arrays.fill(minutesOptAttendees...)
}
```

dmosc replied:
True lol.

anniefu commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -15,9 +15,89 @@
+            availableTimeRanges;
+    }
+
+    private ArrayList<TimeRange> getTimeWindows(Integer[] minutes) {
```
I would add a comment explaining the function's purpose and the intended inputs.

It's ambiguous without looking at the implementation if it's going to return windows where `minutes` is `0` or `1`, and if that represents busy or free.

anniefu commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -15,9 +15,89 @@
+        ArrayList<TimeRange> timeWindowsOptAttendees = getTimeWindows(minutesOptAttendees); // 1440
+
+        long duration = request.getDuration();
+        availableTimeRanges = getAvailableTimeRanges(timeWindowsOptAttendees, duration);
```
You can probably avoid this extra pass by passing the `duration` into `getTimeWindows` and checking `windowSize` is larger than the duration before adding it to the returned `ArrayList<TimeRange>`.

dmosc replied:
Yeah, makes sense. Probably I was trying to achieve some modularity and logic encapsulation since that's healthy, I guess. But I agree, this definitely would reduce time and save lines of code.

anniefu commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -15,9 +15,89 @@
+
+        // If no time range works with optional attendees return the options without them.
+        return availableTimeRanges.size() == 0 ? 
+            getAvailableTimeRanges(timeWindowsNoOptAttendees, duration) : 
```
Probably worth refactoring a bit to avoid unnecessarily computing `timeWindowsNoOptAttendees` In the case that `availableTimeRanges.size() != 0`.

anniefu replied:
Oh, just for clarification, you didn't have to flip the conditional here to `!=`. This comment actually became irrelevant when you removed the `timeWindowsNoOptAttendees = getTimeWindows(minutesNoOptAttendees)`.

anniefu commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -15,9 +15,113 @@
+* 
+* Algorithmic analysis
+* N = Scheduled events
+* M = Attendees within scheduled events
```
I think M is worst case "the entire pool of people who can attend events", because of Collections.disjoint() iterating over both input sets, and being called for every event.

anniefu commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -15,9 +15,113 @@
+        ArrayList<TimeRange> availableTimeRanges = new ArrayList<>();
+        ArrayList<Event> eventsToSet = new ArrayList<>(events);
+
+        Set<String> attendees = new HashSet<String>(request.getAttendees());
```
This copy constructor is sneakily foiling your space complexity analysis. The `new HashSet<String>=(request.GetAttendees())` will allocate new memory. But you can just avoid this by doing `Set<String> attendees = request.getAttendees();` since I don't think you modify the contents anywhere.

ssmall replied:
+1 to Annie's comment, with one additional suggestion: if you do not need to modify `attendees` in the subsequent code, I recommend using [`Collections.unmodifiableSet`](https://docs.oracle.com/javase/8/docs/api/java/util/Collections.html#unmodifiableSet-java.util.Set-) -- it will not copy the underlying set, but will cause an exception if any code tries to modify it.

dmosc replied:
Sweet. Although, I will be switching from _Set_ to a _Collection_ structure to store both **attendees** and **optionalAttendees** since the `request.getAttendees()` return a Collection instead of a Set structure, and that was throwing me this error:

> incompatible types: java.util.Collection<java.lang.String> cannot be converted to java.util.Set<java.lang.String> 

But apart from that, it's working like a charm :)

anniefu reviewed (commented):
Nice! Overall, your code is looking clean and concise. 

I think you're good to go with this algorithm, but just something to ponder: Do you think it's possible to do better than `O(NM)` algorithmic complexity?

anniefu commented:
Just a small note on: *** No breaking changes *** in description. It's great to be aware of that, but it might be a little superfluous, because that's assumed to be true by default, since breaking changes are typically a bad thing and are much rarer to see.

It might be better to invert the pattern omit *** No breaking changes *** in favor of *** breaking changes *** for PRs with breaking changes.

ssmall commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/Event.java:
```diff
@@ -18,6 +18,7 @@
 import java.util.Collections;
 import java.util.HashSet;
 import java.util.Set;
+import java.util.Comparator;
```
nit: unnecessary import

ssmall commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -15,9 +15,113 @@
+* N = Scheduled events
+* M = Attendees within scheduled events
+*
+* Time complexity: N * M
```
Nit: `O(N*M)`

ssmall commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -15,9 +15,113 @@
+* M = Attendees within scheduled events
+*
+* Time complexity: N * M
+* Space complexity: N being the availableTimeRanges ArrayList since it is technically possible to have N available time ranges.
```
nit: `O(N)`

ssmall commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -15,9 +15,113 @@
+* Space complexity: N being the availableTimeRanges ArrayList since it is technically possible to have N available time ranges.
+*   This depends on the analysis rules that are set since all the linear structures
+*   that are used inside the method are for storing and parsing the incoming parameters.
+*   Thus, they shouldn't be considered as extra space because it's not space needed directly
+*   for the solution of the algorithm.
```
nit: I'm confused about what this means. If they aren't necessary for the algorithm, can they be removed?

dmosc replied:
Yeah, this comment was referring to the additional linear structures storing the attendees and optionalAttendees. Which is of course extra space, but I wasn't considering that as part of the space complexity. 

But with the `Collections.unmodifiableCollection` that extra space is not required anymore.

ssmall commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -15,9 +15,113 @@
+import java.util.Set;
+import java.util.HashSet;
+
+/**
+* The FindMeetingQuery receives a new Meeting request to fit inside a
+* calendar day with existing meetings. It provides an interface to
+* handle incoming meetings that must be scheduled for users without
+* generating interference/overlapping with their existing meetings.
+* 
+* Algorithmic analysis
+* N = Scheduled events
+* M = Attendees within scheduled events
+*
+* Time complexity: N * M
+* Space complexity: N being the availableTimeRanges ArrayList since it is technically possible to have N available time ranges.
+*   This depends on the analysis rules that are set since all the linear structures
+*   that are used inside the method are for storing and parsing the incoming parameters.
+*   Thus, they shouldn't be considered as extra space because it's not space needed directly
+*   for the solution of the algorithm.
+*   The arrays minutesNoOptAttendees and minutesOptAttendees can be considered constant since
+*   they represent an abstraction of a whole day divided in minutes.
+*
+* @param Collection<Event> List of scheduled events.
+* @param MeetingRequest Event to fit in the available time windows.
+* @return Collection<TimeRange>
+*/
```
Thank you for providing the detailed description of the algorithmic complexity! This looks great :)

ssmall commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -15,9 +15,113 @@
+        for (int i = 0; i < minutes.length; i++) {
+            if (minutes[i] == 1) {
+                ++windowSize;
+            } else {
+                if (windowSize != 0) {
```
nit: this can just be `} else if (windowSize != 0) {`

ssmall commented on walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:
```diff
@@ -223,6 +286,33 @@ public void justEnoughRoom() {
+    // Leave a small 30 minute window to fit a 15 minute event
+    // that can fit both optional and mandatory attendees.
+    //
+    // Opt Events:            |-C-|
```
Nit: should be `B` instead of `C`

ssmall reviewed (commented):
This is looking really good, Oscar! I concur with all of Annie's comments/suggestions and have a couple more of my own.

dmosc committed:
nitfix (Resolve comments) Resolve nitfix comments.

ssmall reviewed (approved):
LGTM!
